Converting ISoil to AD #229

sveerara opened the pull request:
- Code changes to convert ISoil to use automatic differentiation. 
- Added copies of all tests with AD and the solutions match with the ones without AD. Once we have sufficient confidence on the AD version, we can deprecate the nonAD version or use that only with explicit.
- I have added a jacobian test using a simple 2 element problem quasi-static problem for which the Jacobian entries are close to 1. The relative error with respect to the FD jacobian is within 5e-8. 

Closes #229

moosebuild commented:
Job [Test](https://civet.inl.gov/job/372651/) on 7e3c1e5 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/mastodon/docs/PRs/230/site)

This comment will be updated on new commits.

cbolisetti commented:
Thanks a lot for doing this @sveerara. I will ask @lindsayad to review this if he can since I don't know much about AD.

cbolisetti commented on doc/content/source/materials/ADComputeISoilStress.md:
```diff
@@ -0,0 +1,13 @@
+
+## Description
+
+This material object computes the stress of an I-Soil element and automatically calculates the Jacobian for the element.
```
Can you please add a link to the ISoil documentation here?

lindsayad commented on include/materials/ADComputeISoilStress.h:
```diff
@@ -0,0 +1,162 @@
+  std::vector<std::vector<ADReal>> _youngs;
+
+  /// The coupled variable field providing the soil layer id.
+  const VariableValue & _soil_layer_variable;
```
I guess this doesn't have to be an `ADVariableValue`?

sveerara replied:
This is a constant value which is used as the index to a material property vector and it is only used to determine the appropriate the material properties corresponding to each element. This variable is not used in the residual calculation.

lindsayad replied:
👍

lindsayad commented on src/materials/ADComputeISoilStress.C:
```diff
@@ -0,0 +1,642 @@
+    _layer_ids(getParam<std::vector<unsigned int>>("layer_ids")),
+    _wave_speed_calculation(getParam<bool>("wave_speed_calculation")),
+    _poissons_ratio(getParam<std::vector<Real>>("poissons_ratio")),
+    _density(_wave_speed_calculation ? &getMaterialProperty<Real>("density") : NULL),
```
Do we know that `density` is always going to be independent of the our non-linear degrees of freedom?

lindsayad replied:
`nullptr` is preferable to `NULL`

sveerara replied:
Yes, the density is a material property that is provided by the user and it is independent of nonlinear variables for all the problems we solve using MASTODON. I will change NULL to nullptr

lindsayad commented on src/materials/ADComputeISoilStress.C:
```diff
@@ -0,0 +1,642 @@
+    _pressure_dependency(getParam<bool>("pressure_dependency")),
+    _strength_pressure_correction(1.0),
+    _stiffness_pressure_correction(1.0),
+    _shear_wave_speed(_wave_speed_calculation ? &declareProperty<Real>("shear_wave_speed") : NULL),
```
Same question about independence from nonlinear dofs. This question applies to any place I see declareProperty vs declareADProperty and getMaterialProperty vs getADMaterialProperty

sveerara replied:
Yes, the shear wave speed and P wave speed are material properties that are computed based on the user provided data. It does not vary with nonlinear dofs.

lindsayad commented on test/tests/materials/isoil/tests:
```diff
@@ -67,6 +115,41 @@
+  [../]
+
+  [./data_file_jac_check]
+    # 1 element problem with ISoilAction and automatic differentiation to check auto-generated GQH stress-strain curves
```
What type of boundary conditions do you have for this test? If you had nodal boundary conditions getting applied to every element node, then all the jacobians would be manually set to 1, and your AD jacobians would not be getting tested. I'm guessing this is not the case, but I want to make sure. In general I usually use 2-3 elements so I'm sure I have at least one unconstrained Jacobian entry. Also do you know how many non-linear iterations this test takes? I usually like to get at least two because then that means I'm probably avoiding getting "lucky" in that I might have an initial condition where all variable gradients are zero or something like that.

sveerara replied:
I am using two elements for this test and I am using DirichletBC on the bottom boundary of the bottom element and the top boundary of the top element. So the nodes that are shared between the two elements have free dofs. The iterations look like the one below on most timesteps and I am going through 8 timesteps in this test.

 0 Nonlinear |R| = 2.000501e-03
      0 Linear |R| = 2.000501e-03
      1 Linear |R| = -0.000000e+00
 1 Nonlinear |R| = 9.989995e-04
      0 Linear |R| = 9.989995e-04
      1 Linear |R| = 2.491007e-21
 2 Nonlinear |R| = 4.336809e-19
 Solve Converged!

sveerara commented:
Lot of the Exodiff tests in this material model like GQH and Darendeli tests are checking to make sure that the soil material properties (which are technically inputs to the actual material model) are accurately calculated using the empirical equation that is coded in for these methods. I haven't created Jacobian tests using these auto-generated soil properties because they would result in Jacobian entries with magnitude in the order 10^6. The Jacobian test that I have added tests the main computation involved in the material model. 

There is one other option (pressure dependency) that would affect the Jacobian during runtime based on the nonlinear dof values. In the test that I have added, I have set this option to false. I will try to try to add a jacobian test that includes this capability as well. Thanks for bringing this up.

lindsayad commented:
Alright, let me know when you've added the test and I will approve!!

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/384471/) on 88a055e wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/mastodon/docs/PRs/230/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/mastodon/docs/PRs/230/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format bbcfb87041a8567410745fdebf5b41d27e84d34a`

sveerara commented:
@lindsayad, thanks for approving the PR. I was just about to send you a message saying I have addressed your comments. 

@cbolisetti, as per our discussion, I have added two more tests that check whether the soil loses strength under tension and then gains back strength when it is re-compressed again.
